framework/aster-frame/src/task/atomic.rs

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! # Atomic Execution Mode
+//!
+//! When a CPU core runs in the **atomic mode**,
+//! the kernel code must be executed _atomically_.
+//! Specifically, a kernel panic would be triggered
+//! if such forbidden behaviors are captured from the kernel code in the atomic mode:
+//!
+//! - Any operation that can incur scheduling / context switching:
+//!     - Sleeping;
+//!     - Waiting for I/O;
+//!     - Yielding: yield the processor to other tasks voluntarily;
+//!     - Preemption: get preempted by other tasks;
+//! - And switching to the user space.
+//!
+//! Note that the kernel code is allowed to handle interrupts within the atomic mode.
+//!
+//! The concept of the atomic mode is connected with those of process and interrupt contexts.
+//! A broad classification of execution contexts in a classic monolithic kernel can be:
+//!
+//! | Context Type in Kernel | Process                                 | Interrupt                 |
+//! | ---------------------- | --------------------------------------- | ------------------------- |
+//! | Execution Target       | kernel code on behalf of a user process | interrupt handler         |
+//! | Atomic Requirement     | may switch into or out on demand        | always in the atomic mode |
+//!
+
+use core::{
+    panic,
+    sync::atomic::{AtomicUsize, Ordering::Relaxed},
+};
+extern crate alloc;
+use alloc::{format, string::String};
+
+crate::cpu_local! {
+    static ATOMIC_MODE: AtomicMode = AtomicMode::new(AtomicModeInner::new());
+}
+
+#[derive(Debug, Default)]
+struct AtomicModeInner {
+    /// The number of locks and IRQ-disabled contexts held by the current CPU.
+    cnt: AtomicUsize,
+}
+
+impl AtomicModeInner {
```

Comment:
Actually, I don't get a sense of why this inner type is needed.

Some reasons:
 - The inner type is the only member of `AtomicMode`.
 - All its methods (`inc_cnt`/`dec_cnt`/`cnt`) consist of only one line.
 - All its methods (`inc_cnt`/`dec_cnt`/`cnt`) are called at most twice.
 - The implementation of `AtomicMode` is simple enough, and wrapping the counter in another inner type actually makes the whole codebase _more complex_ (instead of simpler).

With the above reasons in mind, I would suggest just removing this inner type.